src/shared/vscode/commands2.ts

```diff
@@ -0,0 +1,330 @@
+     */
+    build(...args: Parameters<T>): Builder
+
+    /**
+     * Executes the command.
+     *
+     * Only commands registered via {@link Commands} have certain guarantees such as
+     * logging and error-handling.
+     */
+    execute(...parameters: Parameters<T>): Promise<ReturnType<T> | undefined>
+}
+
+interface RegisteredCommand<T extends Callback = Callback> extends Command<T> {
+    dispose(): void
+}
+
+interface DeclaredCommand<T extends Callback = Callback, U extends any[] = any> extends Command<T> {
+    register(...dependencies: U): RegisteredCommand<T>
+}
+
+/**
+ * Minimal wrapper around VS Code's `commands` API to give structure around commands registered
+ * and consumed by the extension.
+ */
+export class Commands {
+    private readonly resources: Map<string, CommandResource<() => unknown, any>> = new Map()
+
+    public constructor(private readonly commands = vscode.commands) {}
+
+    /**
+     * Returns a {@link Command} if the ID is currently registered within VS Code.
+     */
+    public async get(id: string): Promise<Command | undefined> {
+        const registeredCommands = await this.commands.getCommands()
+
+        if (registeredCommands.includes(id)) {
+            const throwOnRegister = () => {
+                throw new Error(`Command "${id}" cannot be re-registered`)
+            }
+
+            return new CommandResource<Callback>({ id, factory: throwOnRegister }, this.commands)
+        }
+    }
+
+    /**
+     * Registers a new command with the VS Code API.
+     */
+    public register<T extends Callback>(id: string, callback: T): RegisteredCommand<T> {
+        const resource = new CommandResource({ id, factory: () => callback }, this.commands)
+
+        return this.addResource(resource).register()
+    }
+
+    /**
+     * Declares the _intent_ to register a command.
+     *
+     * Forward declaration adds one level of mis-direction. This allows for explicit annotation of
+     * not just the command signature but also its immediate dependencies.
+     */
+    public declare<T extends Callback, D extends any[]>(
+        id: string | { id: string; name: string },
+        factory: CommandFactory<T, D>
+    ): DeclaredCommand<T, D> {
+        const resource = typeof id === 'string' ? { id, factory } : { ...id, factory }
+
+        return this.addResource(new CommandResource(resource, this.commands))
+    }
+
+    /**
+     * Convenience method to declare commands directly from a class.
+     *
+     * Mostly used when many commands are associated with the same long-lived context.
+     * Functions are currently bound at declaration time for simplicity, though this
+     * may change in the future.
+     *
+     * #### Example
+     * ```ts
+     * class Foo {
+     *     public square(n: number): number {
+     *         return n * n
+     *     }
+     * }
+     *
+     * const square = Commands.instance.from(Foo).declareSquare('aws.square')
+     * await square.register(new Foo()).execute(5)
```

Comment:
when would we want to declare but not (immediately) register? What do we lose if "declare" is a hidden implementation detail, so that the interface is just "register"?

Reply:
It's just a way to expose the function signature while deferring the registration itself (maybe the command needs `vscode.ExtensionContext`). We need to expose the command on module load, and the best way to do that is through forward declaration. I suppose registration could become an implementation detail, then we just rename declare -> register?

While you _could_ manually write out all the signatures, that's also more to keep track of. This is a way to avoid adding more things to `globals` while keeping many of the ergonomic benefits.

Reply:
> I suppose registration could become an implementation detail, then we just rename declare -> register?

That seems like an ergonomics win. If it adds awkwardness or a perf cost then idk.

Reply:
We can hide `declare` entirely right now but it'd probably put us back where we started with dependencies on every single function signature. I definitely want to get rid of it but we're just not there yet.

My temporary solution was to make it easy to declare commands from a class. That way commands can be treated as a "bundle" and registered all at once.

Reply:
Ok we can skip this for now.